src/util/file_watch.c

```diff
@@ -0,0 +1,331 @@
+            /* Note: this will fire if the modification time changes into the past
+             * as well as the future.
+             */
+            DEBUG(SSSDBG_CRIT_FAILURE, "File [%s] changed\n", fw_ctx->filename);
+            cb->modified = file_stat.st_mtime;
+
+            /* Tell the monitor to signal the children */
+            fw_ctx->fn(fw_ctx->filename, fw_ctx->fn_arg);
+        }
+    }
+
+    ret = create_poll_timer(fw_ctx);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_FATAL_FAILURE,
+              "Error: File [%s] no longer monitored for changes!\n",
+              fw_ctx->filename);
+    }
+}
+
+
+static int watched_file_inotify_cb(const char *filename,
+                                  uint32_t flags,
+                                  void *pvt)
+{
+    struct file_watch_ctx *fw_ctx;
+
+    DEBUG(SSSDBG_TRACE_INTERNAL,
+          "Received inotify notification for %s\n", filename);
+
+    fw_ctx = talloc_get_type(pvt, struct file_watch_ctx);
+    if (fw_ctx == NULL) {
+        return EINVAL;
+    }
+
+    fw_ctx->fn(fw_ctx->filename, fw_ctx->fn_arg);
+
+    return EOK;
+}
+
+
+static int try_inotify(struct file_watch_ctx *fw_ctx)
+{
+#ifdef HAVE_INOTIFY
+    struct snotify_ctx *snctx;
+    /* We will queue the file for update in one second.
+     * This way, if there is a script writing to the file
+     * repeatedly, we won't be attempting to update multiple
+     * times.
+     */
+    struct timeval delay = { .tv_sec = 1, .tv_usec = 0 };
+
+    snctx = snotify_create(fw_ctx, fw_ctx->ev, SNOTIFY_WATCH_DIR,
+                           fw_ctx->filename, &delay,
+                           IN_DELETE_SELF | IN_CLOSE_WRITE | IN_MOVE_SELF | \
+                           IN_CREATE | IN_MOVED_TO | IN_IGNORED,
+                           watched_file_inotify_cb, fw_ctx);
+    if (snctx == NULL) {
+        return EIO;
+    }
+
+    return EOK;
+#else
+    return EINVAL;
+#endif /* HAVE_INOTIFY */
+}
+
+static errno_t fw_watch_file_fallback(TALLOC_CTX *parent_ctx,
+                                      struct file_watch_ctx *fw_ctx)
+{
+    struct file_watch_callback *cb = NULL;
+    struct stat file_stat;
+    int ret, err;
+
+    ret = stat(fw_ctx->filename, &file_stat);
+    if (ret < 0) {
+        err = errno;
+        if (err == ENOENT) {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "File [%s] is missing. Will try again later.\n",
+                  fw_ctx->filename);
+        } else {
+            DEBUG(SSSDBG_FATAL_FAILURE,
+                  "Could not stat file [%s]. Error [%d:%s]\n",
+                  fw_ctx->filename, err, strerror(err));
+        }
+        return err;
+    }
+
+    fw_ctx->poll_check.parent_ctx = parent_ctx;
+
+    cb = talloc_zero(fw_ctx, struct file_watch_callback);
+    if (!cb) {
+        talloc_free(fw_ctx);
+        return ENOMEM;
+    }
+    cb->modified = file_stat.st_mtime;
+
+    DLIST_ADD(fw_ctx->poll_check.callbacks, cb);
```

Comment:
Why do we need a list?

Reply:
Or to attach several callback to a single file?

Reply:
Could be. In the way it was used, it seems it was the first option but, as I said above, I was not taken into account in the caller functions.

Currently each file has it's own context, so it is useless to have a list here.